src/librustc/ty/mod.rs

```diff
@@ -526,6 +526,18 @@ impl<'tcx> TyS<'tcx> {
             _ => true,
         }
     }
+
+    // Returns true if the construction of `self` would require a value of type `!`
+    // to have been constructed. This check is conservative.
+    pub fn requires_never_value(&self) -> bool {
+        match self.sty {
+            ty::TyNever => true,
+            ty::TyRawPtr(ty_and_mut) |
+            ty::TyRef(_, ty_and_mut) => ty_and_mut.ty.requires_never_value(),
```

Comment:
This is actually not as conservative as might be needed. =) For example, it is certainly wrong that `*mut !` requires a never value -- it could legally be null. Similarly, there is some dispute about whether `&!` is truly uninhabited, since in unsafe code such types can come to be (but never get dereferenced). 

I would be happier if we just used `TyNever` here -- but perhaps the real question is why we don't delegate to one of the existing "is uninhabitable" functions.  Was this a deliberate decision, or simply a matter of not knowing about said functions? :)

Reply:
Ah, I shouldn't have forgotten about the weaker unsafe guarantees.
Not using an existing function was just an oversight! Would it be more appropriate to use `conservative_is_uninhabited` or `is_ty_uninhabited_from_all_modules`?